src/gurt/telemetry.c

```diff
@@ -699,6 +742,91 @@ d_tm_build_path(char **path, char *item, va_list args)
 	return rc;
 }
 
+/**
+ * Compute gauge statistics, min, max, mean, and standard deviation
+ * Uses B. P. Welford's algorithm for computing a running variance.
+ * This implementation computes an integer min and max with floating point
+ * mean and standard deviation.
+ *
+ * \param[in]	node		Pointer to a gauge node
+ */
+void
+d_tm_compute_gauge_stats(struct d_tm_node_t *node)
+{
+	double mean;
+	double s;
+	double variance;
+	uint64_t value;
+
+	value = node->dtn_metric->dtm_data.value;
+
+	node->dtn_metric->dtm_stats->k++;
+
+	mean = node->dtn_metric->dtm_stats->mean +
+	       ((value - node->dtn_metric->dtm_stats->mean) /
+	       node->dtn_metric->dtm_stats->k);
+
+	s = node->dtn_metric->dtm_stats->s +
+	    ((value - node->dtn_metric->dtm_stats->mean) * (value - mean));
+
+	node->dtn_metric->dtm_stats->mean = mean;
+	node->dtn_metric->dtm_stats->s = s;
+
+	if (node->dtn_metric->dtm_stats->k > 2) {
+		variance = s / (node->dtn_metric->dtm_stats->k - 1);
+		node->dtn_metric->dtm_stats->std_dev = sqrtl(variance);
+	}
+
+	if (value > node->dtn_metric->dtm_stats->dtm_max.max_int)
+		node->dtn_metric->dtm_stats->dtm_max.max_int = value;
+
+	if (value < node->dtn_metric->dtm_stats->dtm_min.min_int)
+		node->dtn_metric->dtm_stats->dtm_min.min_int = value;
+}
+
+/**
+ * Compute duration statistics, min, max, mean, and standard deviation
+ * Uses B. P. Welford's algorithm for computing a running variance.
+ * This implementation computes a floating point min, max, mean and standard
+ * deviation.
+ *
+ * \param[in]	node		Pointer to a duration node
+ */
```

Comment:
I broke it out into separate functions to keep it simple to follow each one, rather than deal with each data type separately inline.  As with the printing function, if there is consensus that we'd rather have it all in one, I can do that.

Reply:
I have a stronger argument here than for the printing function. Here the logic is nontrivial, and if there's a bug in one of the implementations, you'll need to remember to fix it in both places. Since macros don't care about types, it might be possible to avoid doing anything to explicitly juggle between the types.

Reply:
Using macros to compute this got a little interesting, because even though the macros don't care about types, I still have to differentiate between dtm_min_float and dtm_min_int, dtm_max_float and dtm_max_int.  I made another implementation that uses macros to compute it, but using macros for this purpose seems unnecessarily fragile.  Once I push it, you can see what you think.